scripts/flash.py

```diff
@@ -154,114 +277,249 @@ def __init__(self, link, flash_type):
       self.n_sectors = M25_N_SECTORS
       self.restricted_sectors = M25_RESTRICTED_SECTORS
     else:
-      raise ValueError("flash_type must be \"STM\" or \"M25\"")
+      raise ValueError("flash_type must be \"STM\" or \"M25\", got \"%s\"" \
+                       % flash_type)
+
+  def ihx_n_ops(self, ihx, erase=True):
+    """
+    Find the number of sent SBP messages (erase, program, read) self.write_ihx
+    will require to write a particular intelhex.IntelHex for this instance's
+    flash type.
+
+    Parameters
+    ----------
+    ihx : intelhex.Intelhex
+      intelhex.IntelHex to be written.
+    erase : bool
+      Include number of erase operations required in total.
+
+    Returns
+    -------
+    out : int
+      Number of sent SBP messages (erase, program, read) self.write_ihx will
+      require to write ihx.
+    """
+    return ihx_n_ops(ihx, self.addr_sector_map, erase)
+
+  def inc_n_queued_ops(self):
+    """
+    Increment the count of queued flash operation SBP messages in the STM's
+    flash in a thread safe way.
+    """
+    self.nqo_lock.acquire()
+    self._n_queued_ops += 1
+    self.nqo_lock.release()
+
+  def dec_n_queued_ops(self):
+    """
+    Decrement the count of queued flash operation SBP messages in the STM's
+    flash in a thread safe way.
+    """
+    self.nqo_lock.acquire()
+    self._n_queued_ops -= 1
+    self.nqo_lock.release()
+
+  def get_n_queued_ops(self):
+    """
+    Get the current count of queued flash operation SBP messages.
+
+    Returns
+    -------
+    out : int
+      Get the current count of queued flash operation SBP messages.
+    """
+    return self._n_queued_ops
 
   def __del__(self):
```

Comment:
Its not recommended to put clean up in `__del__` in python. Consider using contexts instead:

http://stackoverflow.com/questions/865115/how-do-i-correctly-clean-up-a-python-object